Added tests for Regexes using regXwild and Conari

3F opened the pull request:
Method                              |         Mean |       Error |      StdDev |       Median 
------------------------------------|--------------|-------------|-------------|--------------
Regex1                              |   9,964.7 ns |    42.90 ns |    38.03 ns |   9,967.4 ns 
Regex2Compiled                      |   4,036.1 ns |    27.87 ns |    24.71 ns |   4,028.9 ns 
Regex3                              |   9,928.7 ns |    50.62 ns |    44.87 ns |   9,926.6 ns 
Regex4                              |   4,606.4 ns |    39.53 ns |    36.98 ns |   4,599.8 ns 
Regex5                              |   1,774.9 ns |    12.26 ns |    11.47 ns |   1,776.1 ns 
Regex5MatchSuccess                  |   1,861.6 ns |    11.82 ns |    11.05 ns |   1,860.7 ns 
Regex6CultureInvariant              |   1,770.9 ns |     9.70 ns |     9.07 ns |   1,774.6 ns 
Regex7                              |   1,777.1 ns |    10.40 ns |     8.68 ns |   1,780.1 ns 
Regex8                              |   1,965.2 ns |    37.30 ns |    36.64 ns |   1,964.5 ns 
Regex9                              |  27,638.2 ns |   594.20 ns | 1,009.00 ns |  27,095.6 ns 
RegXwildExtLambdaConariL            |   1,343.1 ns |     4.81 ns |     4.50 ns |   1,342.3 ns 
RegXwildExtDllImport                |   1,611.4 ns |     7.95 ns |     6.64 ns |   1,612.8 ns 
RegXwildEssLambdaConariL            |   1,770.4 ns |    15.08 ns |    11.77 ns |   1,772.1 ns 
RegXwildEssDllImport                |   2,053.8 ns |     5.92 ns |     5.25 ns |   2,054.3 ns 
RegXwildMatchLambdaConariL          |   1,741.1 ns |     7.54 ns |     6.29 ns |   1,743.1 ns 
RegXwildMatchFastDllImport          |   1,693.0 ns |    10.63 ns |     9.42 ns |   1,694.2 ns 
RegXwildMatchFastLambdaConariL      |   1,506.2 ns |     5.65 ns |     5.28 ns |   1,507.4 ns 
RegXwildExtFastDllImport            |   1,212.2 ns |     7.65 ns |     6.78 ns |   1,212.4 ns 
RegXwildExtFastLambdaConariL        |   1,083.9 ns |     6.45 ns |     6.03 ns |   1,085.7 ns 
PrepareOnce_RegexNonCompiled        |   4,964.5 ns |    25.28 ns |    23.65 ns |   4,962.4 ns 
PrepareOnce_RegexCompiled           | 128,166.4 ns |   837.10 ns |   783.03 ns | 128,256.5 ns 
PrepareOnce_RegexCompiledInvariant  | 124,633.1 ns |   546.17 ns |   510.88 ns | 124,676.8 ns 
PrepareOnce_Conari                  |     436.4 ns |     1.21 ns |     1.07 ns |     436.6 ns 
PrepareOnce_DLL_LoadingAndUnloading | 229,167.7 ns | 1,094.90 ns |   914.29 ns | 229,245.1 ns 
PrepareOnce_ConariFastDllImport     |     665.6 ns |     3.03 ns |     2.68 ns |     666.1 ns 
PrepareOnce_ConariFast              |     395.3 ns |     0.66 ns |     0.62 ns |     395.4 ns

KirillOsenkov commented:
Thanks for the PR!

I'm a bit hesitant about adding a submodule, so I'd prefer to wait until the NuGet package stabilizes to 1.5, no pressure.

Also from my quick reading I think creating and deallocating an unmanaged string should be counted towards the cost of the method, because in my scenario I have a million strings, and I need to match each one, so this isn't something I could do once. It's a bit disappointing but perhaps unavoidable since the engine itself is native.

For now I've pushed your branch to my repo so if anyone is interested they can try experiment with this in this branch:
https://github.com/KirillOsenkov/Benchmarks/tree/regXwild

3F commented:
> I'm a bit hesitant about adding a submodule, so I'd prefer to wait until the NuGet package stabilizes to 1.5, no pressure.

:) This is a regXwild submodule not a Conari.

I also didn't want to add a submodule but I added a bit of changes https://github.com/3F/regXwild/commit/6a7e23c1b12c125bc31806cf42a93e5e1aca5d17 
specially for additional testing by the same rules when the regex patterns has been initialized once without changing in the future.

But I just noticed my thought there was about constants for TXT (according to my other old project) but not for its patterns. Whops.

> Also from my quick reading I think creating and deallocating an unmanaged string should be counted towards the cost of the method, because in my scenario I have a million strings, and I need to match each one, so this isn't something I could do once. It's a bit disappointing but perhaps unavoidable since the engine itself is native.

In future Conari 1.5 I already implemented a *fast native c-strings* without reallocating (allocating a new unmanaged memory regions).

```csharp
using NativeString<CharPtr> data = new("Hello world!");
l.replace<bool>(data, filter, replacement); //regXwild

// data == Hello!
// data == Hi!
// data == Hi, it's me!
// ...
// there are no new allocations in memory
```

Here's result

```csharp
|        Prepare_ConariFastAllocInitTXT | 301.2 ns | 1.78 ns | 1.66 ns | << initial buffer according to text
|     Prepare_ConariFastAllocCharPtrTXT | 590.8 ns | 2.46 ns | 2.31 ns | 

|    Prepare_ConariFastAllocInitPattern | 301.2 ns | 2.10 ns | 1.96 ns | << initial buffer according to pattern
| Prepare_ConariFastAllocCharPtrPattern | 266.7 ns | 1.37 ns | 1.28 ns |

|  Prepare_ConariAllocDeallocCharPtrTXT | 812.7 ns | 6.67 ns | 5.92 ns | - text; FULL alloc/free
|  Prepare_ConariAllocDeallocCharPtrWP1 | 484.4 ns | 3.44 ns | 3.22 ns | - pattern; FULL alloc/free

1 ns   : 1 Nanosecond (0.000000001 sec)
```

So, yes, regXwild was initially created for a native unmanaged C++ projects but not for dotnet.

But even through peinvoke regXwild can be faster than compiled regex patterns (C++ built-in regex engine, or .NET regex engine) as we can see here. 

Say something 2,776.3 ns < less than 126,404.0 ns  ( *(395.3 + 484.4 + 812.7+ 1,083.9)*  < *(1,770.9 + 124,633.1)* )

But the import of DLL can also be costly and for managed dotnet we can, unfortunately, only improve its peinvoke communication. And more like Conari a bit faster helps with this instead of DllImport.

Maybe someday I'll try to re-implement this algorithm in managed dotnet :) Or maybe someone else https://github.com/3F/regXwild/blob/ac1f4780931b310505b2f8bf8a5cdec8a88ee587/regXwild/core/ESS/AlgorithmEss.cpp#L97

> For now I've pushed your branch to my repo so if anyone is interested they can try experiment with this in this branch:

Good! Thanks.

In future Conari 1.6/2.0 versions (I hope it will) obviously there is a need to look at DLL_LoadingAndUnloading optimizations but I think this is possible only through some hack for loadlibrary / some direct processing since I'm sure almost all ns eaten by system calls. Need to think.

3F commented:
> :) This is a regXwild submodule not a Conari.

\*For this PR was enough old Conari version. In the upcoming Conari 1.5, a new types, native chains, helpers, and so on, for the most easy working in one-touch at everything (Memory, Dll, ..). But this will not greatly affect the speed for benchmarks.

3F commented:
I noticed speed in loading of modules was fixed/improved by something together with 1.5. But didn't specifically improve anything in this part. More like it was a bug.

Anyway, nothing surprising when marshaling of the each additional arguments eats a new extra portion of cpu time.
But this is a good reason to think about less arguments in regXwild's peinvoke API as possible for future releases. Well, this is a good reason to rewrite logic using dotnet but that's another story.

Perhaps the best solution for the cases like this is to push 1 raw data argument onto the stack then additionally evaluate this on native c++ manually. In any case `func(TCHAR*, TCHAR*, flagcfg_t, MatchResult*)` definitely will be slower than `funcNameAs0Nullptr(TCHAR*, TCHAR*)`

3F commented:
Oh, I am also puzzled about speed for marshaled Unicode characters since managed strings in dotnet CLR are UTF-16 and there's no reason to be slower than multibyte processing *except* everything that I said above, well...
